File: langLibs/PHP/prflr.php
<?

/*
 *  HOW TO USE 
 * 
 * // configure profiler
 * // set  profiler server:port  and  set source for timers 
 * PRFLR::init(192.168.1.45-testApp', 'yourApiKey');
 * 
 * 
 * //start timer
 * PRFLR::Begin('mongoDB.save');
 * 
 * //some code
 * sleep(1);
 * 
 * //stop timer
 * PRFLR::End('mongoDB.save');
 * 
 */

class PRFLR {

    private static $sender;

    public static function init($source, $apikey) {
        self::$sender = new PRFLRSender();

        if (!self::$sender->apikey = $apikey)
            throw new Exception('Unknown apikey.');

        if (!self::$sender->socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP))
            throw new Exception('Can\'t open socket.');

        if (!$source)
            self::$sender->source = isset($_SERVER['SERVER_ADDR']) ? $_SERVER['SERVER_ADDR'] : 'Unknown';
        else
            self::$sender->source = $source;
        self::$sender->thread = uniqid();
    }

    public static function begin($timer) {
        self::$sender->Begin($timer);
    }

    public static function end($timer, $info = '') {
        self::$sender->End($timer, $info);
    }

    public function __destruct() {
        unset(self::$sender);
    }

}

class PRFLRSender {

    private $timers;
    public $socket;
    public $delayedSend = false;
    public $source;
    public $thread;
    public $ip;
    public $port = 4000;
    public $apikey;

    public function __construct() {
        $this->ip = gethostbyname("prflr.org");
    }

    public function __destruct() {
        socket_close($this->socket);
    }

    public function Begin($timer) {
        $this->timers[$timer] = microtime(true);
    }

    public function End($timer, $info = '') {

        if (!isset($this->timers[$timer]))
            return false;

        $time = round(( microtime(true) - $this->timers[$timer] ) * 1000, 3);

        $this->send($timer, $time, $info);

        unset($this->timers[$timer]);
    }

    private function send($timer, $time, $info = '') {

        // format the message
        $message = join(array(
            substr($this->thread, 0, 32),
            substr($this->source, 0, 32),
            substr($timer, 0, 48),
            $time,
            substr($info, 0, 32),
            substr($this->apikey, 0, 32),

                ), '|');

        if ($this->socket) {
            socket_sendto($this->socket, $message, strlen($message), 0, $this->ip, $this->port);
        } else {
            throw new Exception("Socket not exist\n");
        }
    }

}

File: langLibs/Python/prflr.py
#!/usr/bin/env python

# Threadsafe api
# usage example:

# from prflr import Prflr
# p = Prflr('key', '192.168.1.45-testApp')
# p.begin('mongoDB.save')
# # do something usefull
# p.end('mongoDB.save', 'results')

import sys
import time
import socket
import threading

# True if we are running on Python 3.
PY3 = sys.version_info[0] == 3

class Prflr:
    def __init__(self, api_key, source=''):
        self.addr = ('prflr.org', 4000)
        self.source = source if source else socket.gethostname()
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.key = api_key
        self.timers = {}
        self.lock = threading.RLock()

    def begin(self, timer):
        thread_id = threading.current_thread().ident
        with self.lock:
            if not self.timers.get(thread_id):
                self.timers[thread_id] = {}
            self.timers[thread_id][timer] = time.time()

    def end(self, timer, info=''):
        thread_id = threading.current_thread().ident
        timers = self.timers.get(thread_id)
        if not timers:
            return False
        prev_time = timers.get(timer)
        if not prev_time:
            return False
        spent = round((time.time() - prev_time) * 1000, 3)
        self.send(timer, spent, info, thread_id)
        with self.lock:
            timers.pop(timer)
        return True

    def send(self, timer, time, info, thread_id):
        msg = '|'.join([str(thread_id)[:32],
                        self.source[:32],
                        timer[:48],
                        str(time)[:16],
                        info[:32],
                        self.key[:32]])
        if PY3:
            self.socket.sendto(bytes(msg, 'utf-8'), self.addr)
        else:
            self.socket.sendto(msg, self.addr)
